pkg/registry/core/service/allocator/storage/storage.go

```diff
@@ -85,6 +85,12 @@ func (e *Etcd) Allocate(offset int) (bool, error) {
 	e.lock.Lock()
 	defer e.lock.Unlock()
 
+	//lint:ignore SA9003 we try to sync etcd and the port allocation data
+	if err := e.sync(); err != nil {
```

Comment:
syncing before every allocate seems like it could impact performance quite negatively. do we actually need to sync before allocate, or can we only sync in case of failure to see if the port we're being asked to allocate was actually freed in the meantime?

Reply:
yeah, I thought about that, but is not that what the `tryUpdate()` method is trying to do just 5 lines below?
https://github.com/kubernetes/kubernetes/pull/86527/files#diff-b80626acaf6cac8561b8be49df2214adR99
If that's the case, we should just remove the block of code that tries to allocate out of the `tryUpdate`method, right?
I totally agree on the performance impact, I've assigned wojtek hoping he can help or give some guidance.

Reply:
It's not the most frequently used function, but I agree with Jordan that we should try to avoid that.

That said, I think I don't really understand the code below. For me calling `Allocate` once and then calling it inside tryUpdate doesn't make much sense.
Unless I'm missing something, I think we should remove the first Allocate completely.

Reply:
That indeed works and solves the problem, but I'm afraid of doing any change since all methods apply the same logic

https://github.com/kubernetes/kubernetes/blob/b1e86d3cb308b65ded26892d71723187a71b6e7b/pkg/registry/core/service/allocator/storage/storage.go#L118-L119

https://github.com/kubernetes/kubernetes/blob/b1e86d3cb308b65ded26892d71723187a71b6e7b/pkg/registry/core/service/allocator/storage/storage.go#L149-L150

Should we remove all the lines trying to do the local operation first?

Reply:
I've updated the PR avoiding any operation locally first

Reply:
Removing the lines doesn't work, I think that the code in `tryUpdate` overwrites etcd directly :/

@liggitt do you know if is possible to use a watcher on the etcd key to update the local bitmap on changes?
I was checking the code and this is using the apiserver Storage interface https://godoc.org/k8s.io/apiserver/pkg/storage that allow to use Watchers, but no idea on if it's possible to use it here or how 🤷‍♂